In [1]:
import numpy as np
import pandas as pd

In [2]:
data = pd.DataFrame({
    'child': ['A', 'B', 'C', 'D', 'E'],
    'age': [5, 6, 7, 8, 9],
    'height': [100, 105, 108, 112, 115]
})

data

child  age  height
0     A    5     100
1     B    6     105
2     C    7     108
3     D    8     112
4     E    9     115

In [3]:
age = data['age']
height = data['height']

mean_age = np.mean(age)
mean_height = np.mean(height)

print(f'mean_age = {mean_age}')
print(f'mean_height = {mean_height}')

denominator = np.sum((age - mean_age)**2)
numerator = np.sum((age - mean_age) * (height - mean_height))

print(f'numerator = {numerator}')
print(f'denominator = {denominator}')

mean_age = 7.0
mean_height = 108.0
numerator = 37.0
denominator = 10.0


In [4]:
m = numerator / denominator
b = mean_height - (m * mean_age)
m
b

np.float64(3.7)

np.float64(82.1)

Final Equation:$$Height = 3.7(Age) + 82.1$$

Prediction

For Age 10: $y = 3.7(10) + 82.1 =$ $119.1 \text{ cm}$

In [5]:
x = 10
y = m * x + b
y

np.float64(119.1)

In [6]:
for x in age:
  y = m * x + b
  print(y)

100.6
104.3
108.0
111.69999999999999
115.4


In [7]:
age
type(age)
height
type(height)

0    5
1    6
2    7
3    8
4    9
Name: age, dtype: int64

pandas.core.series.Series

0    100
1    105
2    108
3    112
4    115
Name: height, dtype: int64

pandas.core.series.Series

In [8]:
height_pred = m * age + b
height_pred

0    100.6
1    104.3
2    108.0
3    111.7
4    115.4
Name: age, dtype: float64

In [9]:
# Sum Squared Residual
ssr = np.sum((height - height_pred) ** 2)
ssr

np.float64(1.1000000000000085)

In [10]:
#  Mean Squared Error
np.sum((height - height_pred) ** 2) / len(age)

np.float64(0.2200000000000017)

In [11]:
# Root Mean Squared Error
np.sqrt((np.sum((height - height_pred) ** 2)) / len(age))

np.float64(0.46904157598234475)

Formula:$$R^2 = 1 - \frac{SS_{res}}{SS_{tot}}$$Where:

$SS_{res}$ (Residual Sum of Squares): $\sum (y_{true} - \hat{y})^2$ — The error our model makes.

$SS_{tot}$ (Total Sum of Squares): $\sum (y_{true} - \bar{y})^2$ — The variation in the data itself.

In [12]:
ss_res = np.sum((height - height_pred) ** 2)
ss_res
ss_tot = np.sum((height - mean_height) ** 2)
ss_tot

r2 = 1 - (ss_res / ss_tot)
r2

np.float64(1.1000000000000085)

np.float64(138.0)

np.float64(0.9920289855072463)

### 2. Multiple linear regression (with just 1 feature)

`X` should be a matrix (two dimensions)

In [13]:
age = np.array([5, 6, 7, 8, 9])
height = np.array([100, 105, 108, 112, 115])

In [14]:
np.expand_dims(age, 1)
np.expand_dims(age, 1).shape

array([[5],
       [6],
       [7],
       [8],
       [9]])

(5, 1)

In [15]:
X = age.reshape(-1, 1)
X
X.shape

array([[5],
       [6],
       [7],
       [8],
       [9]])

(5, 1)

In [16]:
bias_col = np.ones(len(age))
bias_col

array([1., 1., 1., 1., 1.])

In [17]:
X = np.c_[bias_col, X]
X
X.shape

array([[1., 5.],
       [1., 6.],
       [1., 7.],
       [1., 8.],
       [1., 9.]])

(5, 2)

In [18]:
y = height.reshape(-1, 1)
y
y.shape

array([[100],
       [105],
       [108],
       [112],
       [115]])

(5, 1)

The Transpose ($X^T$)

Shape: $(2 \times 5)$

In [19]:
X_tr = X.T
X_tr.shape

(2, 5)

The Gram Matrix ($X^T X$)

In [20]:
X_tr.shape, X.shape

((2, 5), (5, 2))

Matrix Multiplication

For matrices $A \in \mathbb{R}^{m \times n}$ and $B \in \mathbb{R}^{n \times p}$, the elements of the product $C = AB$ are given by:$$C_{ij} = \sum_{k=1}^{n} A_{ik} B_{kj}$$

Explanation:

$A$ is an $m \times n$ matrix (rows $\times$ columns).

$B$ is an $n \times p$ matrix.

The resulting matrix $C$ is $m \times p$.

To find the value at row $i$, column $j$ of the result, you perform a dot product of the $i$-th row of $A$ and the $j$-th column of $B$.

In [21]:
C = np.zeros(shape=(2, 2))
C

for i in range(2):
  for j in range(2):
    C[i, j] = np.dot(X_tr[i], X[:, j])

C

array([[0., 0.],
       [0., 0.]])

array([[  5.,  35.],
       [ 35., 255.]])

 The Gram Matrix ($X^T X$)
 
 - This is the "Sum of Squares" matrix.
 
- Action: Multiply $X^T \cdot X$.Formula 

Hint:$$\begin{bmatrix} \text{Count}(n) & \sum x \\ \sum x & \sum x^2 \end{bmatrix}$$

In [22]:
gram_matrix = X_tr @ X
gram_matrix
gram_matrix.shape

array([[  5.,  35.],
       [ 35., 255.]])

(2, 2)

The Moment Vector ($X^T y$)This represents the correlation between features and target.
- Action: Multiply $X^T \cdot y$.

Formula Hint:

$$\begin{bmatrix} \sum y \\ \sum (x \cdot y) \end{bmatrix}$$

In [23]:
moment_vector = X_tr @ y
moment_vector

array([[ 540.],
       [3817.]])

The Inverse ($(X^T X)^{-1}$)This is the hardest manual step.

You need the inverse of the $2 \times 2$ matrix from Party B.

Formula for $2 \times 2$ Inverse:$$A^{-1} = \frac{1}{ad - bc} \begin{bmatrix} d & -b \\ -c & a \end{bmatrix}$$

Calculation:

- Find Determinant ($ad - bc$): $(5 \cdot 255) - (35 \cdot 35)$.
- Swap $a$ and $d$, change signs of $b$ and $c$.
- Divide by determinant.

In [24]:
np.linalg.inv(X_tr @ X)

array([[ 5.1, -0.7],
       [-0.7,  0.1]])